### PR TITLE
Inconsistency in PreTrainedModel.resize_token_embeddings When ZeRO3 Is Enabled

### PR DESCRIPTION
**Pull request by sinamoeini**
# What does this PR do?

This PR addresses https://github.com/huggingface/transformers/issues/25241.

In previous implementation when ZeRO stage 3 was enbaled, resize_token_embeddings would create independent PyTorch weights on each device. Here we ensure that new embeddings are created with DeepSpeed init, and are properly partitioned accros devices.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

- @pacman100
- @ydshieh

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_utils.py
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1483]: would like @pacman100 and @sgugger to take a look here.

File: src/transformers/modeling_utils.py
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1482]: I am not familiar on this topic, but why this line is removed? I am sure there is good reason.

**Review by ydshieh — COMMENTED** [review]
Overall LGTM, but it's better for @pacman100 and the core maintainer to take a look, especially for the DeepSpeed part of code.

Thank you for your work!

File: src/transformers/modeling_utils.py
Comment (shady273185Gunit) [reply src/transformers/modeling_utils.py:1474]: Ek smaak n kus koop en jou daar in sit

File: src/transformers/modeling_utils.py
Comment (sinamoeini) [reply src/transformers/modeling_utils.py:1482]: I think github diff is a bit misleading here. These two lines were replaced with (see above)
```
            params = [old_embeddings.weight, new_embeddings.weight]
            with deepspeed.zero.GatheredParameters(params, modifier_rank=0):
```
The only thing  I removed was 
```
 if torch.distributed.get_rank() == 0:
```
since it is a parallel process it is not saving latency. I can add it back in if you think it would make more sense.

File: src/transformers/modeling_utils.py
Comment (sinamoeini) [reply src/transformers/modeling_utils.py:1474]: Will add them back in, after other reviewers give me their feedbacks

**Review by sgugger — APPROVED** [review]
This looks good to me, but like the expert @pacman100 to chime in too.
Thanks for your PR!

**Comment by sinamoeini** [comment]
@pacman100 would you mind reviewing this PR?

**Review by pacman100 — APPROVED** [review]
Thank you @sinamoeini for the fixes, LGTM! 🤗

**Comment by sinamoeini** [comment]
Thank you @sgugger would you mind merging this?

File: src/transformers/modeling_utils.py
Comment (ydshieh) [reply src/transformers/modeling_utils.py:1474]: Before merge, don't forget this comment 🙏

**Comment by ydshieh** [comment]
There is a comment to be addressed please 🙏

**Comment by sinamoeini** [comment]
@ydshieh I addressed your comment however the documentation build cancelled. Could you kick it off again and merge if there are no more issues? thanks

**Comment by ydshieh** [comment]
It's fine. Thanks again 🚀 !

**Pull request merged**
